File: rust/sieve_atkin.rs
static N: uint = 100000000;
fn main() {
  stieve_atkin();
}

fn stieve_atkin() {
  let nsqrt: uint = ((N as float).sqrt() + 1.0) as uint;
  let mut is_prime = ~[false, ..N];
  let mut n: uint;
  let mut primes = ~[];

  for x in range(1, nsqrt) {
    for y in range(1, nsqrt) {
      n = 4 * (x*x) + (y*y);
      if n <= N && (n % 12 == 1 || n % 12 == 5) { is_prime[n] = !is_prime[n]; }

      n = 3 * (x*x) + y*y;
      if n <= N && n % 12 == 7 { is_prime[n] = !is_prime[n]; }

      n = 3 * (x*x) - y*y;
      if x > y && n <= N && n % 12 == 11 { is_prime[n] = !is_prime[n]; }
    }
  }

  for x in range(5, nsqrt) {
    if is_prime[x] {
      let mut y = x*x;
      while y < N {
        is_prime[y] = false;
        y += x*x;
      }
    }
  }

  is_prime[2] = true;
  is_prime[3] = true;

  for x in range(0, N) { if is_prime[x] { primes.push(x) } }
}

File: ruby/sieve_atkin.rb
N = 100000000
is_prime = []
primes = []
nsqrt = Math::sqrt(N).floor

(1..nsqrt).each do |x|
  (1..nsqrt).each do |y|
    n = 4 * (x * x) + y * y
    is_prime[n] = !is_prime[n] if n <= N && (n % 12 == 1 || n % 12 == 5)

    n = 3 * (x * x) + y * y
    is_prime[n] = !is_prime[n] if n <= N && n % 12 == 7

    n = 3 * (x * x) - y * y
    is_prime[n] = !is_prime[n] if x > y && n <= N && n % 12 == 11
  end
end

(5..nsqrt).each do |x|
  next if !is_prime[x]

  y = x * x
  (y..N-1).step(y).each { |i| is_prime[i] = false }
end

is_prime[2] = true
is_prime[3] = true

is_prime.each_with_index { |prime, i| primes << i if prime }
puts primes.count
